Initial implementation

stsewd opened the pull request:
The extension itself is pretty simple, most of the files are docs and tests.

I'm not 100% sure about the extension name/settings.

agjohnson commented on docs/configuration.rst:
```diff
@@ -0,0 +1,48 @@
+.. confval:: sharedconf_docsets
+
+   A dictionary defining the projects (docsets) that will share the configuration.
+   You must define at least one docset.
```
I've described this as "docset" as well, but perhaps "projects' is the nomenclature that most folks understand. Probably worth reflecting that in the configuration naming too if we want to clarify

agjohnson commented on docs/implementation.rst:
```diff
@@ -0,0 +1,7 @@
+Implementation
+==============
+
+Basically the extension changes the
```
```suggestion
Basically, this extension changes the
```

agjohnson commented on docs/implementation.rst:
```diff
@@ -0,0 +1,7 @@
+
+Basically the extension changes the
+:py:attr:`app.srcdir <sphinx:sphinx.application.Sphinx.srcdir>`
+value and configuration options at runtime at the earliest possible
+step in the build process.
```
```suggestion
value and configuration options during runtime, at the earliest possible
step in the build process.
```

agjohnson commented on docs/index.rst:
```diff
@@ -0,0 +1,29 @@
+
+Share a single ``conf.py`` file between several Sphinx projects,
+useful if you want to import several projects from the same
+repository on `Read the Docs <https://readthedocs.org/>`__.
```
I'd have the same feedback from the readme here

agjohnson commented on tests/test_extension.py:
```diff
@@ -0,0 +1,156 @@
+
+class TestExtension:
+    def _read_text(self, path):
+        """Compatible read_text for Sphinx<3."""
```
One point for dropping Sphinx 2.

agjohnson commented on sharedconf/extension.py:
```diff
@@ -0,0 +1,85 @@
+    # Taken from
+    # https://github.com/sphinx-doc/sphinx/blob/586de7300a5269de93b4d813408ffdbf10ce2409/sphinx/config.py#L222-L222
+    pre_init_options = {"needs_sphinx", "suppress_warnings", "language", "locale_dirs"}
+    for option, value in docset_config.items():
```
This method of overriding the configuration file seems redundant -- the user would already have independent values in `docs/project1/conf.py` and `docs/project2/conf.py`. Could we get away with doing an import of most of these values, instead of duplicating this data in a common config file?

I'm sure users will just want to do a:

```python

import projecta.conf as projecta_conf
import projectb.conf as projectb_conf

sharedconf_docsets = {
    "projecta": {"config": {
        "project": proejcta.project,
        "version": projecta.version,
        ...
    }
}

stsewd replied:
I like the idea of having the settings in different files and also have the possibility of using just one file (if the user only needs to override a couple of values on each project).

I see this being used in different ways:

```python
from multiproject.utils import get_project

extensions = [
    "sharedconf.extension",
    "sphinx.ext.intersphinx",
    "sphinx.ext.autodoc",
]

master_doc = "index"
language = "en"

multiproject_projects = {
    "api": {},
    "dev": {
        "config": {
            # Is this way of setting the options still useful?
            # Could be for simple projects,
            # so users don't need to import stuff.
            "project": "Dev project",
        },
    },
    "user": {},
    'auto': {
        # This is almost the same as using `config`,
        # but we set the values automatically from the `auto/conf.py` file?.
        # This has the same restrictions as using `config`.
        'config_file': 'auto/conf.py',
    },
}

current_project  = get_project(multiproject_projects)

# Set all values directly
# -----------------------

if current_project == 'api':
    from api.conf import *
elif current_project == 'user':
    from user.conf import *

# Set value by value
# ------------------

import api.conf as api_conf
import user.conf as user_conf
if current_project == 'api':
    config = api_conf
elif current_project == 'user':
    config =  user_conf

# Replace the original values
project = config.project
version = config.version
# Extending the original value
extensions += config.extensions
```

agjohnson replied:
> and also have the possibility of using just one file

Ahh I see where you're coming from. We could certainly support both. I do think the auto import method would be used the most, so I suppose I'm just advocating for adding that now/soon.

My guess would be that users would still maintain 2 full sphinx subprojects for more standard local development, so could see the `multiproject_projects['dev']` being an advanced use case.

I did like how your original example had all the configuration that a project would need in one place, instead of requiring manual `import` from the user. Manual set up could be an advanced usage, but we would recommend something like your `multiproject_projects['auto']` for almost all use cases.

agjohnson commented on setup.cfg:
```diff
@@ -0,0 +1,27 @@
+[metadata]
+name = sphinx-sharedconf
+version = attr: sharedconf.__version__
+author = Read the Docs
```
```suggestion
author = Read the Docs, Inc
```

agjohnson commented on setup.cfg:
```diff
@@ -0,0 +1,27 @@
+version = attr: sharedconf.__version__
+author = Read the Docs
+author_email = support@readthedocs.org
+description = Share a single conf.py file between several Sphinx projects
```
Description edits from readme here as well.

agjohnson reviewed (commented):
Very thorough, I appreciate the addition of some test cases on this code :+1: 

Major points I raised were the naming/description of the package and some of the concepts. I also hadn't thought much about the functionality to assist conditionally setting configuration values. This seems helpful, but it also seems like something we could derive from the existing subprojects instead of redefining.

stsewd commented on .circleci/config.yml:
```diff
@@ -0,0 +1,60 @@
+    steps:
+      - run-tox:
+          version: py310
+          sphinx-version: 42,43,44,latest
```
was about to ask about that, what should be our position about what versions of sphinx we should support.

stsewd replied:
> This is also just a small extension though, so it's also not a big deal to oversubscribe here.

Like test for the latest version of 3.x and 4.x only?

agjohnson replied:
Good question. We haven't had this specific discussion yet. We've discussed it on the theme side, but have much stricter constraints there and need to support more than we really want to.

I suppose this is mostly to help put upward pressure on user's Sphinx dependencies. Trying to get more users off old Sphinx and on 3 or 4 would be a small benefit. At the end of the day, supporting 2 major releases of Sphinx already seems fair enough. I don't feel strongly here though, but if you agree, let's prune!

stsewd commented on .readthedocs.yaml:
```diff
@@ -0,0 +1,16 @@
+
+python:
+  install:
+    - requirements: docs/requirements.txt
```
do you mean using extras_require? `pip install foo[docs]`?

agjohnson replied:
That's a good pattern to test, but really any other pattern is helpful to use to get some different perspective. I've used poetry on a few repos too. It's not important for this immediate PR, and we can always play around with things at a later time. I just like to advocate against `requirements.txt` though, it's sort of a last ditch method for dependencies these days.

stsewd commented on README.md:
```diff
@@ -1,2 +1,14 @@
+
+Share a single `conf.py` file between several Sphinx projects.
+useful if you want to import several projects from the same
+repository on [Read the Docs](https://readthedocs.org).
```
I like `sphinx-readthedocs-multiproject` :D

stsewd replied:
but it's also kind of long, so maybe sphinx-multiproject is okay?

agjohnson replied:
heh yeah, that is quite long. `sphinx-multiproject` seems like a good plan :+1: 

I could see us eventually maybe having a small collection of extensions, and maybe we have a `sphinx-readthedocs` at that point. This seems like it would be a good fit in that umbrella, as you're right that this is probably only really applicable to projects on RTD.

stsewd commented:
I have updated the PR:

- Rename extension to sphinx-multiproject
- I have renamed the concept of "docset" to "project"
- Read the values from {path}/conf.py by default (can be disabled)
- I'm still keeping the "config" option on each project definition (but I'm fine removing it if you consider so)
- Only test the latest release of sphinx 3.x, and all mayor releases of 4.x.

agjohnson commented on docs/configuration.rst:
```diff
@@ -0,0 +1,91 @@
+
+   .. confval:: path
+
+      By default it's the same as the name of the project.
+      It can be a relative (to the conf.py file) or an absolute path.
```
```suggestion
      Path to the nested project's ``conf.py`` file.
      By default, the key name used in ``multiproject_projects`` is also the directory name for the nested project's ``conf.py`` file.
      This can be a relative path (to the main ``conf.py`` file) or an absolute path.
```

agjohnson commented on docs/configuration.rst:
```diff
@@ -0,0 +1,91 @@
+   .. confval:: config
+
+      A dictionary of options specific to this project.
+      Useful if you don't want to have a specific ``conf.py`` file
```
```suggestion
      This is useful if you don't want to have a specific ``conf.py`` file
```

agjohnson commented on docs/configuration.rst:
```diff
@@ -0,0 +1,91 @@
+
+   .. confval:: use_config_file
+
+      If `True` the config from ``{path}/conf.py`` would be included for this project.
```
```suggestion
      If ``True``, the configuration values from ``{path}/conf.py`` would be automatically imported when building this project.
```

agjohnson commented on docs/quickstart.rst:
```diff
@@ -0,0 +1,91 @@
+   # File: docs/conf.py
+
+   extensions = [
+      "multiproject.extension",
```
Perhaps this is a pattern to reign in. We did this historically on a few extensions, but most modern Sphinx extension just use the top level package name:

```suggestion
      "multiproject",
```

This is annoying to change later, so might be worth doing now, unless there is a particular reason we need the nested namespace.

stsewd replied:
This was so it wont give an import error at https://github.com/readthedocs/sphinx-multiproject/blob/b82ae09ecda055148ae9be13b6326534a6e75a95/multiproject/__init__.py#L1-L1 when installing the extension (since sphinx isn't installed yet). But looks like when not using a `setup.py` (but a setup.cfg) the file isn't evaluated

https://github.com/readthedocs/sphinx-multiproject/blob/b82ae09ecda055148ae9be13b6326534a6e75a95/setup.cfg#L3-L3

Did a quick test and looks like shouldn't be a problem :)

agjohnson replied:
Oh nice, I didn't know that was possible!

I've been liking bump2version for this as well, and avoiding all of the import issues on `__version__` entirely (plus versions strings like in package.json/etc)

agjohnson commented on multiproject/__init__.py:
```diff
@@ -0,0 +1 @@
+__version__ = "0.1.0rc"
```
We can also go straight to 1.0rc1 here. We usually start semver wrong, but 0.x is for initial development. We are arguable skipping initial development, and going straight to a stable public API. This would indicate a 1.0 release.

Anyways, just making sure we don't release a 0.1.0 as the first release. `0.1.0rc` sounds like our first release is 0.1.0

```suggestion
__version__ = "1.0.0rc1"
```

agjohnson commented on multiproject/extension.py:
```diff
@@ -0,0 +1,104 @@
+        config_file = Path(app.srcdir) / CONFIG_FILENAME
+        if config_file.exists():
+            project_config.update(
+                eval_config_file(
```
:+1: I was originally describing directly doing an `import`, but this is a much better version of that.

agjohnson reviewed (approved):
Looks great! I noted some copy change to add a little clarity to usage, but I mostly understood how to configure the extension either way.

Do we have a strong reason to use `multiproject.extension` as the extension import name? I believe we do this on our sphinx rtd extension, but it's not necessary and might make sense to match what other extensions do.

agjohnson commented:
Changes look reasonable! I jumped in to get this merged today. We can revisit making version strings nice or automated, seems the solution still brought on cyclical imports.
